### PR TITLE
Add experimental path to use transformer dialect only in the codegen

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Add a new option so that we can run transformer dialect after tile and
distribute on workgroup. This allows bypassing the problem of setting
the number of workgroups. For now we just pass a tile size through
command line.
This will then allow writing transform dialect pipelines decided by
KernelConfig. Later the transform dialect can evolve to do the tile and
distribute to workgroup.

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
This is a little bit of a hack. If you really want to go down this path, thats fine, but this is probably very short term....

**Comment by ThomasRaoux** [comment]
> This is a little bit of a hack. If you really want to go down this path, thats fine, but this is probably very short term....

Yes, definitely a hack. That would allow me to get the first step to have some transformer based codegen we can use by default though. Once we can replace TileAndDistributeToWorkgroup by a transformer based conversion we can remove the flag.
What do you think? Do you see a better way?

**Comment by ThomasRaoux** [comment]
> Then even the first level tile + distribute (with interchange, etc.) can be handled through the transform dialect. If you want this for short term evaluation thats one thing, but this needs to be fixed and if someone can help with it, that would be great.

Yes that makes sense, I was proposing that mainly to workaround the discussion about computing the number of workgroup causing problems. I believe Nicolas has plan to be able to use foreach_thread at the workgroup level along with the transform dialect, that would remove the need for this workaround. 
I'll let Nicolas decides what he thinks is the best next step. I don't have a strong opinion as long as there is a way for me to use transforms for codegen only.

**Comment by nicolasvasilache** [comment]
Let's have **something** that we can iterate on.

Atm the alternative is https://github.com/iree-org/iree/pull/10131 which is still flagged as having issues.

> Then even the first level tile + distribute (with interchange, etc.) can be handled through the transform dialect.
This is what https://github.com/iree-org/iree/pull/10131 does but the `materialize the code for workgroup count calculation` is still missing. I do not have a good comprehension of the assumptions of the linked code and how things connect to values passed to `stream.dispatch`.

At this point I consider that these 2 PRs narrow down the problem significantly (i.e. this does not have tiling happening at the Flow level) and we'll need @MaheshRavishankar to dig in to bring this home and spell the Flow / Transform Dialect Codegen boundary the way he thinks it should be.

**Review by nicolasvasilache — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
> Let's have **something** that we can iterate on.
> 
> Atm the alternative is #10131 which is still flagged as having issues.
> 
> > Then even the first level tile + distribute (with interchange, etc.) can be handled through the transform dialect.
> 
> This is what #10131 does but the `materialize the code for workgroup count calculation` is still missing. I do not have a good comprehension of the assumptions of the linked code and how things connect to values passed to `stream.dispatch`.
> 
> At this point I consider that these 2 PRs narrow down the problem significantly (i.e. this does not have tiling happening at the Flow level) and we'll need @MaheshRavishankar to dig in to bring this home and spell the Flow / Transform Dialect Codegen boundary the way he thinks it should be.

Ok, if this is on my plate, I'd rather get this out of the way sooner rather than later... I havent blocked this PR, neither have I blocked the other PR (which really I didnt even get a chance to review before it landed, but thats secondary).. Personally I am having trouble following the cadence of things here since they are various different things being done in ways I cant fully connect to what the end state being worked towards is. I'd really appreciate if I am given a starting point to start digging into this, it will just be much faster for me.  If the pointer is just "start from the test" then that's what I will do from next week.
Just to clarify, there is no grand vision on the boundary between Flow/Transform DIalect/Codegen boundary. Anything within `hal.executable.variant` is within what can be part of Transform dialect. Anyone working in IREE should be able to connect the pieces. Like everyone here, I am over-subscribed too. I was trying to avoid dropping everything else (which also need to land) to prioritize this, but if that needs to be done, then I can look into it next week.

**Pull request merged**
